### PR TITLE
Create activity for showing recent transactions.

### PR DESCRIPTION
**Pull request by Ishadijcks**
Added a button on the main activity to this page, and a back button to return to the main page.
Created 2 small xml files to display the transactions neater.

Images:
https://cloud.githubusercontent.com/assets/9715314/26077443/52ef08b8-39bc-11e7-8175-34af13d3b043.png
https://cloud.githubusercontent.com/assets/9715314/26077440/52ec74ea-39bc-11e7-9eb0-b31ca7a40ec6.png
https://cloud.githubusercontent.com/assets/9715314/26077441/52edbfbc-39bc-11e7-8e46-baed01395891.png
https://cloud.githubusercontent.com/assets/9715314/26077442/52ee5eea-39bc-11e7-91fe-4844d043af69.png

### CODE REVIEW

File: BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/RecentTransactionsActivity.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+     */
+    public void showRecentTransactions(List<? extends Transaction> transactionList) {
+        listDataHeader = new ArrayList<>();
+        listDataChild = new HashMap<>();
+
+        expListView = (ExpandableListView) findViewById(R.id.recentTransactionsList);
+        listAdapter = new ExpandableListAdapter(this, listDataHeader, listDataChild);
+        expListView.setAdapter(listAdapter);
+
+        if (transactionList != null) {
```
Comment (joskuijpers) [inline BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/RecentTransactionsActivity.java]: You can do instead:
```java
if (transactionList == null) { showError(); return}
if size == 0 { show error; return }

rest of code
```

Moves the code a bit to the left.

File: BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/RecentTransactionsActivity.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+//                        details.add(transactionList.get(i).getDate().toString());
+//                    }
+
+                    if (transactionList.get(i).getCounterparty() != null) {
+                        details.add("" + transactionList.get(i).getCounterparty().toString());
+                    }
+                    listDataChild.put(listDataHeader.get(i), details);
+                }
+            } else {
+                updateErrorDisplay("No transactions have been made yet.");
```
Comment (joskuijpers) [inline BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/RecentTransactionsActivity.java]: The "No items" view is not an 'error'

File: BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/RecentTransactionsActivity.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+    /*
+    Updates the error display if the recent transactions could not be retrieved.
+     */
+    public TextView updateErrorDisplay(String message) {
+        TextView textView = (TextView) findViewById(R.id.recentTransactionsError);
+        textView.setText(message);
+        textView.setTextSize(20);
+        return textView;
+
```
Comment (joskuijpers) [inline BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/RecentTransactionsActivity.java]: Unnecessary whitespace

File: BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/ExpandableListAdapter.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
+package nl.tudelft.ewi.ds.bankchain.activities;
+
+/**
+ * Created by TUDelft SID on 15-5-2017.
```
Comment (joskuijpers) [inline BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/ExpandableListAdapter.java]: TUDelft SID?

File: BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/ExpandableListAdapter.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+public class ExpandableListAdapter extends BaseExpandableListAdapter {
+
+    private Context _context;
+    private List<String> _listDataHeader; // header titles
+    // child data in format of header title, child title
+    private HashMap<String, List<String>> _listDataChild;
+
+    public ExpandableListAdapter(Context context, List<String> listDataHeader,
+                                 HashMap<String, List<String>> listChildData) {
+        this._context = context;
```
Comment (joskuijpers) [inline BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/ExpandableListAdapter.java]: Why the underscores? We do that nowhere ;)

File: BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/ExpandableListAdapter.java [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+    @Override
+    public View getChildView(int groupPosition, final int childPosition,
+                             boolean isLastChild, View convertView, ViewGroup parent) {
+
+        final String childText = (String) getChild(groupPosition, childPosition);
+
+        if (convertView == null) {
+            LayoutInflater infalInflater = (LayoutInflater) this._context
```
Comment (joskuijpers) [inline BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/ExpandableListAdapter.java]: No need to use "this." here. You can just use `_context` (or `context`)

File: BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/ExpandableListAdapter.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+        TextView txtListChild = (TextView) convertView
+                .findViewById(R.id.lblListItem);
+
+        txtListChild.setText(childText);
+        return convertView;
+    }
+
+    @Override
+    public int getChildrenCount(int groupPosition) {
+        return this._listDataChild.get(this._listDataHeader.get(groupPosition))
```
Comment (joskuijpers) [inline BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/ExpandableListAdapter.java]: Same with the this._ here, of course. And down below

File: BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/MainActivity.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+import nl.tudelft.ewi.ds.bankchain.R;
+
+public class MainActivity extends AppCompatActivity {
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_main);
+
```
Comment (joskuijpers) [inline BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/MainActivity.java]: Whitespace

**Comment by Ishadijcks** [comment]
Header seems to be missing on the main screen.

**Comment by Ishadijcks** [comment]
Fixed all previous mentioned issues

File: BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/ExpandableListAdapter.java [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+
+        TextView lblListHeader = (TextView) convertView
+                .findViewById(R.id.lblListHeader);
+        lblListHeader.setTypeface(null, Typeface.BOLD);
+        lblListHeader.setText(headerTitle);
+
+        return convertView;
+    }
+
+    @Override
```
Comment (RvdK) [inline BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/ExpandableListAdapter.java:105]: what does this method do?

File: BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/RecentTransactionsActivity.java [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
+        setSupportActionBar(toolbar);
+        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
+        retrieveRecentTransactions();
+    }
+
+
+    public void retrieveRecentTransactions() {
+        Environment v = new Environment();
```
Comment (RvdK) [inline BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/RecentTransactionsActivity.java]: wouldnt't you want to keep this more centralized? it looks like you are creating a new session for every call.

File: BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/RecentTransactionsActivity.java [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+        for (int i = 0; i < transactionList.size(); i++) {
+            listDataHeader.add(transactionList.get(i).getDescription());
+            List<String> details = new ArrayList<>();
+
+            if (transactionList.get(i).getValue() != null && transactionList.get(i).getCurrency() != null) {
+                details.add(transactionList.get(i).getValue().toString() + " " + transactionList.get(i).getCurrency().toString());
+            }
+            if (transactionList.get(i).getDirection() != null) {
+                details.add("" + transactionList.get(i).getDirection().toString());
+            }
```
Comment (RvdK) [inline BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/RecentTransactionsActivity.java:115]: ToDo in pull request. is this not part of the story?

File: BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/RecentTransactionsActivity.java [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+
+    public void retrieveRecentTransactions() {
+        Environment v = new Environment();
+        v.bank = "Bunq";
+        v.url = "";
+        v.apiKey = "";
+
+        Bank b = new BankFactory(v).create();
+
+        b.createSession()
```
Comment (RvdK) [inline BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/RecentTransactionsActivity.java:52]: this call could use some comments about what it does

File: BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/RecentTransactionsActivity.java (hunk 11 above)
Comment (Ishadijcks) [reply BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/RecentTransactionsActivity.java:115]: getDate() doesn't work correctly and throws an exception when called. The story was for the UI to display the information about a transaction, which it does for all available information.
I could remove the TODO and the commented code, but I think it's better to leave it as a reminder and uncomment when getDate() works.

File: BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/ExpandableListAdapter.java (hunk 9 above)
Comment (Ishadijcks) [reply BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/ExpandableListAdapter.java:105]: It's overriden from `ExpandableListAdapter` . It it to indicate that the ids of the objects in the list might change if more items are added.

File: BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/RecentTransactionsActivity.java (hunk 10 above)
Comment (Ishadijcks) [reply BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/RecentTransactionsActivity.java]: Yes, we could create a singleton manager so we don't have to create a new Bank every time the activity is opened. That is however not in the scope of this task.

File: BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/RecentTransactionsActivity.java (hunk 12 above)
Comment (Ishadijcks) [reply BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/RecentTransactionsActivity.java:52]: I did not create this method, merely made it visible in the app. With the proposed refactor from above, more comments have to be added.

File: BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/RecentTransactionsActivity.java (hunk 10 above)
Comment (RvdK) [reply BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/RecentTransactionsActivity.java]: can't you at least keep your bank instance in the class itself?

File: BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/RecentTransactionsActivity.java (hunk 12 above)
Comment (RvdK) [reply BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/RecentTransactionsActivity.java:52]: then put some comments about that that it is placeholder code and will be fixed (and maybe create a story/task for it).

Comment (Ishadijcks) [reply BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/RecentTransactionsActivity.java:52]: Fair enough

File: BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/RecentTransactionsActivity.java (hunk 10 above)
Comment (Ishadijcks) [reply BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/RecentTransactionsActivity.java]: When exiting the recent transaction screen, the entire activity is destroyed. When the activity is reopened, the band instance is created again.

File: BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/RecentTransactionsActivity.java (hunk 12 above)
Comment (Ishadijcks) [reply BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/RecentTransactionsActivity.java:52]: Done

**Review by RvdK — APPROVED** [review]
All coments were adressed
